xcell/cls/cov.py

```diff
@@ -429,39 +431,90 @@ def get_covariance(self):
+                cov_G = self._get_covariance_spin0_approx(cw, s_a1, s_a2, s_b1,
+                                                          s_b2, cla1b1, cla1b2,
+                                                          cla2b1, cla2b2, wa,
+                                                          wb)
 
             else:
-                cov = nmt.gaussian_covariance(cw, s_a1, s_a2, s_b1, s_b2,
-                                              cla1b1, cla1b2, cla2b1, cla2b2,
-                                              wa, wb)
-        else:
-            size1 = self.clA1A2.get_ell_cl()[1].size
-            size2 = self.clB1B2.get_ell_cl()[1].size
-            cov = np.zeros((size1, size2))
+                cov_G = nmt.gaussian_covariance(cw, s_a1, s_a2, s_b1, s_b2,
+                                                cla1b1, cla1b2, cla2b1, cla2b2,
+                                                wa, wb)
+
+        if self.nl_marg and notnull:
+            cov_nlm = self.get_covariance_nl_marg()
 
-        if self.nl_marg:
-            cov += self.get_covariance_nl_marg()
+        if self.m_marg and notnull:
+            cov_mm = self.get_covariance_m_marg()
 
-        if self.m_marg:
-            cov += self.get_covariance_m_marg()
+        if self.do_NG and notnull:
+            fsky = self.data.data['cov'].get('fsky_NG', -1.)
+            if fsky < 0:  # Calculate from masks
+                fsky = np.mean(((m_a1 > 0) & (m_a2 > 0) &
+                                (m_b1 > 0) & (m_b2 > 0)))
+            kinds = self.data.data['cov'].get('NG_terms',
+                                              ['1h'])
+            for kind in kinds:
+                cov_NG += self.get_covariance_ng_halomodel(s_a1, s_a2,
+                                                           s_b1, s_b2,
+                                                           fsky, kind)
 
-        self.cov = cov
-        np.savez_compressed(fname, cov=cov)
+        self.cov = cov_G + cov_nlm + cov_mm + cov_NG
+        np.savez_compressed(fname, cov=self.cov, cov_G=cov_G, cov_NG=cov_NG,
+                            cov_nl_marg=cov_nlm, cov_m_marg=cov_mm)
         self.recompute_cov = False
-        return cov
+        return self.cov
+
+    def get_covariance_ng_halomodel(self, s_a1, s_a2, s_b1, s_b2,
+                                    fsky, kind='1h'):
+        ellA = self.clA1A2.b.get_effective_ells()
+        ellB = self.clB1B2.b.get_effective_ells()
+        nclsa = np.max([1, s_a1 + s_a2])
+        nclsb = np.max([1, s_b1 + s_b2])
+        cov = np.zeros([len(ellA), nclsa, len(ellB), nclsb])
+        th = Theory(self.data.data)
+        mpA1, mpA2 = self.clA1A2.get_mappers()
+        mpB1, mpB2 = self.clB1B2.get_mappers()
+        trlist = self.data.data['tracers']
+        ccl_trA1 = th.compute_tracer_ccl(self.trA1,
+                                         trlist[self.trA1],
+                                         mpA1)
+        bA1 = trlist[self.trA1].get('bias', 1.)
```

Comment:
We should unify the way we get the bias, to avoid doing `.get('bias', 1)` in every place.

Reply:
What do you propose?

Reply:
Maybe a method in `Data`? I don't like it too much, but since it is related to reading the configuration of the tracers, it might not be too odd.

Reply:
Done

Reply:
Great! Thanks!